ST in output not checked

v0d1ch opened the pull request:
fix #596 

## Why

We want to make sure that the ST token is present in the tx output of the head validator and that it has correct `CurrencySymbol` that uniquely identifies a Hydra Head instance. We can't tell if some arbitrary NFT with the valid `TokenName` is indeed our ST token we expect so to tackle this security hole we are enforcing the check for the actual `CurrencySymbol` we expect.

## What

- Add `CurrencySymbol` to the datum and check the tx output for its existence in the head validators check for `collectCom`, `close`  and `contest`.
- Add the same check into commit and initial validator scripts in the `ViaAbort` case
- Add Mutation tests for each of the affected validator checks

Note: This PR removes some of the tests that did exist previously. TxSpec ones are outdated, highy specialized and we have a coverage for the same properties already. It also removes the mutation tests tied to `MutateHeadScriptInput` since we are not using the head script output to check the datum contents.

To check before merging:
* [ ] CHANGELOG is up to date
* [x] Up to date with master

v0d1ch committed:
Start with writing a test

v0d1ch committed:
Refactor mutation in the Abort

abailly committed:
Mutate policy id of PT in commit transaction

v0d1ch committed:
Introduce datum for the intial validator

v0d1ch committed:
Fix failing test

v0d1ch committed:
Remove redundant imports

v0d1ch committed:
Rename checkAuthor

v0d1ch committed:
Remove a note about not checking currency symbol

v0d1ch committed:
Remove redundant import

v0d1ch committed:
Start to thread the head token for Close tx

v0d1ch committed:
Use testPolicyId in Close mutation

v0d1ch committed:
Replace all outputs with different policy id

v0d1ch committed:
Replace only single tokens

v0d1ch committed:
Don't use HeadContext for checkClose - pass the policy in the datum

v0d1ch committed:
Use only necessary data from HeadContext

v0d1ch committed:
Add original policy to replace in replacePolicyIdWith

v0d1ch committed:
Thread head policy id through Closed constructor

v0d1ch committed:
Fix collectCom validator and check head id for mutation to work

v0d1ch committed:
Change the number of outputs for the fanout tx to fix the tests

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -560,8 +560,7 @@ addPTWithQuantity tx quantity =
         -- NOTE: We do not expect Ada or any other assets to be minted, so
-        -- we can take the policy id from the headtake the policy id from
-        -- the head.
+        -- we can take the policy id from the head
```
Nice catch 👀

v0d1ch committed:
Check for head id in contest validator

v0d1ch committed:
Reduce outputs further

v0d1ch committed:
Remove traces in head validator and bring back the old outputs number
for fanout tx

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -388,6 +388,7 @@ forAllFanout action =
             & label ("Fanout size: " <> prettyLength (countAssets $ txOuts' tx))
  where
+  -- TODO: Can we go back to 70 outputs here?
   maxSupported = 70
```
how has removing the traces solved this TODO?

v0d1ch replied:
Removing traces reduced tx sizes so I could revert to the old number of outputs we had.

v0d1ch replied:
I removed the line also

v0d1ch committed:
Remove traces from initial and commit validators

v0d1ch committed:
Remove a TODO

v0d1ch committed:
Rename MutatePolicyId to MutateHeadId

v0d1ch committed:
Add Mutation test for contest txs

v0d1ch committed:
Fix the mutation test for contest txs

v0d1ch committed:
Rename head state head id fields

v0d1ch committed:
Cleanup

v0d1ch committed:
Add ST check for abort validator too

v0d1ch committed:
Remove abort check, ST is burned

v0d1ch committed:
Add comments about ST token being present in output

v0d1ch committed:
Small refactor

v0d1ch committed:
Revert some code in Close mutation tests

v0d1ch committed:
Increase the number of parties

v0d1ch committed:
Bump max parties to 8

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -194,7 +197,7 @@ commitTx scriptRegistry networkId party utxo (initialInput, out, vkh) =
     fst (initialReference scriptRegistry)
   initialDatum =
-    mkScriptDatum $ Initial.datum ()
+    mkScriptDatum $ Initial.InitialDatum (headIdToCurrencySymbol headId)
```
why this is not `mkTxOutDatum`?

v0d1ch replied:
Probably because we are creating a `ScriptWitness` using this datum so the types would not align. Previously we used the helper function `datum` to encode empty datum but now we are using `InitialDatum` with the new field for currency symbol.

ffakenz commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -577,22 +577,23 @@ contestationPeriodParser =
 -- and on-chan validators (see 'computeCollectComCost' 'computeAbortCost')
 maximumNumberOfParties :: Int
-maximumNumberOfParties = 4
+maximumNumberOfParties = 8
```
🚀

ffakenz commented on hydra-plutus/src/Hydra/Contract/HeadState.hs:
```diff
@@ -24,17 +24,20 @@ data State
       , snapshotNumber :: SnapshotNumber
       , utxoHash :: Hash
       , contestationDeadline :: POSIXTime
+      , closedHeadId :: CurrencySymbol
```
if `initialHeadId`,  `openHeadId`, and `closedHeadId` are the same, does it not make sense to name them the same?

v0d1ch replied:
I wanted for us to be able to see in code from which constructor exactly the value comes since it gets passed to different validator checks and could be useful when debugging.

ch1bo replied:
We would know from which constructor this comes from as we are not using field names as functions (see our coding style) and do bind them to variables when we need to distinguish them. So this is not an argument and I'm with franco on this one.

The other fields do also share a name so it's not even consistent.

If you want to make an argument against sharing names between fields of sum types, we should change all of the fields here and also think about doing that everywhere else (after having the discussion on the coding style).

v0d1ch replied:
Ok but look at this piece of code where we have `CloseObservation`, `ClosedState` and `OpenState` all having `headId` field:

```
  let CloseObservation{threadOutput, headId, snapshotNumber} = observation
  guard (headId == headId)
  let ClosedThreadOutput{closedContestationDeadline} = threadOutput
  let event =
        OnCloseTx
          { snapshotNumber
          , contestationDeadline = posixToUTCTime closedContestationDeadline
          }
  let st' =
        ClosedState
          { closedThreadOutput = threadOutput
          , headId = headId
          , closedHeadTokenScript = openHeadTokenScript
          }
  pure (event, st')
 where
  OpenState
    { headId
    , openHeadTokenScript
    } = st

```
I can introduce qualified names but introducing different field names felt natural to me.

ch1bo replied:
I was not taking about the off-chain `State` types, but the on-chain datum type here. This module is currently using non-prefixed names and so this new field should follow that for consistency reasons!

At the moment we are even more inconsistent 
![image](https://user-images.githubusercontent.com/2621189/211750116-b0bc3c8b-d560-4e7c-b302-c3fb7270342b.png)

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -116,7 +119,10 @@ healthyParties = flip generateWith 42 $ do
+  peter <- arbitrary
+  judy <- arbitrary
+  john <- arbitrary
+  pure [alice, bob, carol, peter, judy, john]
```
welcome to the party: peter, judy and john

ffakenz replied:
should we invite 2 more people to the party?

v0d1ch replied:
I added two more

v0d1ch committed:
Include two more parties

pgrange commented on hydra-node/test/Hydra/Chain/Direct/Contract/Abort.hs:
```diff
@@ -168,32 +168,28 @@ genAbortMutation (tx, utxo) =
     , SomeMutation UseInputFromOtherHead <$> do
```
This mutation is older than this pull request.

As we did not check for the the _headId_ before, I would have expected this mutation `UseInputFromOtherHead` to fail on master but it does not so I guess the test did not cover some case that this P.R. is fixing.

So now I expect this mutation to make the test succeed inside the P.R. because of the fix brought by the P.R.

It is my understanding that this P.R. improves this mutation to cover the case that was not covered before but it's not clear to me how just looking at the changes.

Could you explain in a word. or two what was missing in the previous version of the mutation?

v0d1ch replied:
Previously we had a check in the head validator that detects if the ST token is present but we were using just the token name `HydraHeadV1` for this check and we never checked PolicyId/CurrencySymbol of the asset which is what the gap is about. We used to grab whatever CS we got for the token with a matching name and then use that to check the signatures for example. The abort test you are mentioning was properly altering the policy id of the token and we were checking if all the PTs contain the same CS  but we never checked if the CS we get is the one that belongs to that specific head.

ch1bo commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ changes.
 
 ## [0.9.0] - UNRELEASED
 
+- Maximum number of parties is increased from 4 to 8.
```
I don't agree. The benchmarks show an upper bound of 4 for `abort` and `collectCom`?

v0d1ch replied:
yes I just realized how tx-cost executable works. I've removed the line from the changelog.

ch1bo replied:
Still.. why can we increase it to 8? https://github.com/input-output-hk/hydra/pull/658/files#r1061391934

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api/Value.hs:
```diff
@@ -96,3 +96,9 @@ fromPlutusValue plutusValue =
 toPlutusValue =
   Ledger.transValue . toLedgerValue
+
+toPlutusPolicyId :: PolicyId -> CurrencySymbol
```
Although trivial, having a haddock would be good here (IIRC there was some pruning of non-documented functions employed in this package)

v0d1ch replied:
:+1: I also renamed `toPlutusPolicyId` to `toPlutusCurrencySymbol`

ch1bo commented on hydra-plutus/src/Hydra/Contract/Commit.hs:
```diff
@@ -121,8 +121,7 @@ validator (_party, headScriptHash, commit) consumer ScriptContext{scriptContextT
                         Nothing -> True
                         Just Commit{preSerializedOutput} ->
                           -- There should be an output in the transaction corresponding to this preSerializedOutput
-                          traceIfFalse "cannot find committed output" $
```
Why remove this trace altogether?

ch1bo commented on hydra-plutus/src/Hydra/Contract/Initial.hs:
```diff
@@ -38,6 +36,12 @@ import qualified PlutusTx
 import qualified PlutusTx.Builtins as Builtins
 
+newtype InitialDatum = InitialDatum
+  { headPolicyId :: CurrencySymbol
```
You called it `headId` in other places. I would prefer that and it would be less odd given the plutus type is called different.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Initial.hs:
```diff
@@ -64,33 +68,31 @@ type RedeemerType = InitialRedeemer
+  (policyId, ourParticipationTokenName) =
     case AssocMap.toList (getValue initialValue) of
-      [_someAdas, (_headCurrencyHopefully, tokenMap)] ->
+      [_someAdas, (headCurrencyHopefully, tokenMap)] ->
```
I think we should directly (as directly as possible given an associative list) access the policy id we care about using an `AssocMap.lookup` using the expected currency symbol from the datum. This *must* yield a singleton entry with quantity 1 and the token name is the pubkeyhash we expect in `txInfoSignatories`.


Should: refactor this accordingly

ch1bo commented on hydra-plutus/src/Hydra/Contract/Initial.hs:
```diff
@@ -108,11 +110,7 @@ checkCommit commitValidator committedRef context@ScriptContext{scriptContextTxIn
-        traceIfFalse ("initialValue: " `appendString` debugValue initialValue) $
-          traceIfFalse ("comittedValue: " `appendString` debugValue committedValue) $
-            lockedValue == initialValue + committedValue
+    lockedValue == initialValue + committedValue
```
I'm okay with removing the debugging, but having a hint which check failed is still useful. We should keep the `traceIfFalse "lockedValue does not match"` or replace it with a well-known error code (to save sizes).. but this is really pre-mature optimization. We can do this "minification" as a very last step before 1.0.0.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -182,24 +130,21 @@ checkAbort context@ScriptContext{scriptContextTxInfo = txInfo} headContext parti
     && everyoneHasCommitted
-    && mustBeSignedByParticipant context headContext
+    && mustBeSignedByParticipant ctx initialHeadId
+    && hasSTToken initialHeadId outValue
```
It's just called ST. So `hasST`or `hasStateToken` or `hasStateThreadToken` would be better names IMO

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -459,8 +444,8 @@ mustContinueHeadWith ScriptContext{scriptContextTxInfo = txInfo} headAddress cha
-        traceIfFalse "wrong output head datum" (findTxOutDatum txInfo o == datum)
-          && traceIfFalse "wrong output value" (checkOutputValue (xs <> rest))
+        findTxOutDatum txInfo o == datum
+          && checkOutputValue (xs <> rest)
```
These traces where helpful as well!

v0d1ch replied:
I'll try to add them back in. While working on this I experienced a problem with the tx sizes so removed traces to optimize the tx sizes.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -459,8 +444,8 @@ mustContinueHeadWith ScriptContext{scriptContextTxInfo = txInfo} headAddress cha
       traceError "no continuing head output"
     (o : rest)
       | txOutAddress o == headAddress ->
```
Could: inline the `mkHeadAddres` / `headAddress` here. AFAIS it's not used anywhere else and this would make the calls to the `checkXXX` functions simpler.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Commit.hs:
```diff
@@ -121,8 +121,7 @@ validator (_party, headScriptHash, commit) consumer ScriptContext{scriptContextT
+                          preSerializedOutput `elem` (Builtins.serialiseData . toBuiltinData <$> txInfoOutputs txInfo)
                     -- NOTE: In the Collectcom case the inclusion of the committed output 'commit' is
                     -- delegated to the 'CollectCom' script who has more information to do it.
                     ViaCollectCom -> True
```
Must: check also that the ST is consistent with the datum and present in the `collectCom` transaction output (see spec).

ch1bo commented on hydra-plutus/src/Hydra/Contract/Commit.hs:
```diff
@@ -121,8 +121,7 @@ validator (_party, headScriptHash, commit) consumer ScriptContext{scriptContextT
                         Just Commit{preSerializedOutput} ->
                           -- There should be an output in the transaction corresponding to this preSerializedOutput
-                          traceIfFalse "cannot find committed output" $
-                            preSerializedOutput `elem` (Builtins.serialiseData . toBuiltinData <$> txInfoOutputs txInfo)
```
Should: check that the ST is consistent with the datum and getting burned (see spec).

v0d1ch replied:
I understand the part _ST is consistent with the datum_ but how is ST getting burned here? Is this something that is done in the abort?

v0d1ch replied:
Tactical meeting shed some light on this so we are checking for token burning also in the commit validator and can decide later if this is redundant/increasing script size too much.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -176,7 +179,7 @@ commitTx ::
   -- locked by initial script
   (TxIn, TxOut CtxUTxO, Hash PaymentKey) ->
```
FWIW, the `headId` would already be included as the policy id of the PT included in this `TxOut`.

v0d1ch replied:
It is a bit easier to have the value already passed in. Do you think I should refactor to get the headId out of this TxOut instead?

ch1bo replied:
No. Just wanted to point out that info so you are aware :)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -194,7 +197,7 @@ commitTx scriptRegistry networkId party utxo (initialInput, out, vkh) =
     fst (initialReference scriptRegistry)
   initialDatum =
-    mkScriptDatum $ Initial.datum ()
+    mkScriptDatum $ Initial.InitialDatum (headIdToCurrencySymbol headId)
```
Could: Use record syntax if you already introduce a new type with record fields.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -311,8 +322,9 @@ closeTx ::
   PointInTime ->
   -- | Everything needed to spend the Head state-machine output.
   OpenThreadOutput ->
+  HeadId ->
```
Either document all of them or none.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -838,7 +843,7 @@ genCommit =
 genCollectComTx :: Gen (ChainContext, [UTxO], InitialState, Tx)
 genCollectComTx = do
-  ctx <- genHydraContextFor 3
+  ctx <- genHydraContextFor maximumNumberOfParties
```
:+1:

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -577,22 +577,23 @@ contestationPeriodParser =
 -- and on-chan validators (see 'computeCollectComCost' 'computeAbortCost')
 maximumNumberOfParties :: Int
-maximumNumberOfParties = 4
+maximumNumberOfParties = 8
```
The results on CI and locally don't show 8 as maximum but remains at 4?

v0d1ch replied:
Yes, my bad. I just realized we are outputting the limits by running the tx-cost executable

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -572,3 +571,13 @@ addPTWithQuantity tx quantity =
+
+-- | Replace original policy id with the arbitrary one
+replacePolicyIdWith :: PolicyId -> PolicyId -> TxOut a -> TxOut a
+replacePolicyIdWith originalHeadId otherHeadId output =
```
I would not mention "headId" here at all.. it's only about policy ids

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -572,3 +571,13 @@ addPTWithQuantity tx quantity =
+  let value = txOutValue output
+      newValue = valueFromList $ swapPolicyId <$> valueToList value
+      swapPolicyId = \case
+        (AssetId policyId t, q) | policyId == originalHeadId, q == 1 -> (AssetId otherHeadId t, q)
```
Why limit to single token values? This function might be much more useful if it is not restricting it to this exact quantity (also we could use it for burning different tokens!). Furthermore, it's not document that it would behave this way.
```suggestion
        (AssetId policyId t, q) | policyId == originalHeadId -> (AssetId otherHeadId t, q)
```

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -334,7 +335,7 @@ forAllClose ::
   Property
 forAllClose action = do
   -- FIXME: we should not hardcode number of parties but generate it within bounds
```
When you change code, look for comments to update around to fix. This FIXME is now solved.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -143,6 +151,7 @@ data CloseMutation
   | MutateValidityInterval
   | MutateCloseContestationDeadline
   | MutateCloseContestationDeadlineWithZero
+  | MutateHeadId
```
Can you document what exactly this mutation simulates?

It seemingly changes the policy of head tokens in the input & output.

Would a more subtle attack like closing a head by using another Head's PT also be prevented?

v0d1ch replied:
I'll try to write a test for this case.

v0d1ch replied:
We already simulate using a validator with the _wrong_ ST/ coming from a different head

ch1bo replied:
So what is it testing now?

When trying to document this, first explanation I tried to give was
```
  -- | Change the policy id of the state token (ST) to simulate closing this
  -- head using an ST from a different head.
```

But this is not what we actually do, right? We are changing the policy from the value on the spending & producing side and see the validator fail because of the datum. But this is not very realistic (as a previous validator would check that the datum holds the right policy id).

A proper attack here would be to spend another output of the same validator (another "open head") and try to close the head-under-attack using the ST from that other validator. I believe this mutation would be red on `master` and green on this branch.

ch1bo replied:
Another check we should be doing is to try to really mutate the head id .. i.e. change the producing (output) side to be different than the spending (input) side policy id.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -45,31 +46,37 @@ healthyCloseTx =
+    }
+
+headInput :: TxIn
+headInput = generateWith arbitrary 42
```
What's the point of pushing these to the module-level? Some of the names are quite vague and when pushing identifiers out of their semantic context, names quickly become too vague.

v0d1ch replied:
I am using `headResolvedInput` downstream in the same file for mutation tests. `headDatum` and `openThreadOutput` can be put back.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Abort.hs:
```diff
@@ -168,32 +168,28 @@ genAbortMutation (tx, utxo) =
     ]
+
+removePTFromMintedValue :: TxOut CtxUTxO -> Tx -> Value
+removePTFromMintedValue output tx =
```
This function is super cryptic. I see you only moved it out, but it's very hard to read what it's doing :/

v0d1ch replied:
I've tried to improve on that.

ch1bo replied:
It's better, thanks.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -116,7 +119,12 @@ healthyParties = flip generateWith 42 $ do
+  john <- arbitrary
+  mary <- arbitrary
+  tom <- arbitrary
+  pure [alice, bob, carol, peter, judy, john, mary, tom]
```
Why are so many needed to expose a critical behavior of the validator?

v0d1ch replied:
They are not _needed_ just wanted to exercise scenario with more parties.

v0d1ch replied:
After realizing this doesn't add any particular value I am reverting it to what it was - three parties.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/CollectCom.hs:
```diff
@@ -171,7 +179,7 @@ genCollectComMutation (tx, utxo) =
     , SomeMutation MutateHeadTransition <$> do
         changeRedeemer <- ChangeHeadRedeemer <$> (Head.Close 0 . toBuiltin <$> genHash <*> arbitrary)
-        changeDatum <- ChangeHeadDatum <$> (Head.Open <$> arbitrary <*> arbitrary <*> (toBuiltin <$> genHash))
+        changeDatum <- ChangeHeadDatum <$> (Head.Open <$> arbitrary <*> arbitrary <*> (toBuiltin <$> genHash) <*> arbitrary)
```
This test is basically useless.. so many things are arbitrarily wrong that it will rarely highlight a flaw in the validator (it's easy to make a validator error out with garbage data)

v0d1ch replied:
I've changed this to construct the correct headId at least.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Commit.hs:
```diff
@@ -35,33 +37,31 @@ healthyCommitTx =
+commitVerificationKey = generateWith arbitrary 42
+
+initialOutput :: TxOut CtxTx
+initialOutput = mkInitialOutput Fixture.testNetworkId Fixture.testPolicyId commitVerificationKey
```
We usually prefix these top-level values `healthyXXX` to indicate that they comprise the healthy case and are selectively mutated below.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Commit.hs:
```diff
@@ -35,33 +37,31 @@ healthyCommitTx =
       commitParty
       (Just healthyCommittedUTxO)
-      (initialInput, toUTxOContext initialOutput, initialPubKeyHash)
+      (Fixture.testSeedInput, toUTxOContext initialOutput, initialPubKeyHash)
```
I was confused of the `testSeedInput` usage here. This `TxIn` is the one used by the `testPolicyId`, while this input should correspond to a hypothetical UTxO produced by the initial transaction.

Should: keep it arbitrarily generated (as before)

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -50,27 +51,28 @@ healthyContestTx =
+headResolvedInput :: TxOut CtxUTxO
+headResolvedInput =
+  mkHeadOutput testNetworkId testPolicyId headTxOutDatum
+    & addParticipationTokens healthyParties
```
Should: rename this to `healthyHeadTxOut`

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -50,27 +51,28 @@ healthyContestTx =
       }
 
+headTxOutDatum :: TxOutDatum CtxUTxO
+headTxOutDatum = toUTxOContext (mkTxOutDatum healthyClosedState)
```
This is only used in `headResolvedInput`. Keep it as local as possible, especially when names are vague.

ch1bo reviewed (changes requested):
I think we are missing the datum check in the commit validator. Other than that, some minor stuff in the mutations and I don't see where this is now enabling 8 parties to collect/abort a head?

v0d1ch committed:
Use maxNumberOfParties in more tests

v0d1ch committed:
Add haddock and rename toPlutusPolicyId to toPlutusCurrencySymbol

v0d1ch committed:
Add back the trace in the v_commit

v0d1ch committed:
Rename headPolicyId to headId in v_initial

v0d1ch committed:
Add missing trace

v0d1ch committed:
Refactor v_initial token discovery

v0d1ch committed:
Rename hasSTToken to hasST

v0d1ch committed:
CollectCom refactor arguments

v0d1ch committed:
Rename state headId fields to be the same

pgrange committed:
Restore traces

pgrange committed:
JSON format changed

v0d1ch committed:
Rename initialOutput to healthyInitialOutput

v0d1ch committed:
Bring back initialInput

v0d1ch committed:
Keep headTxOutDatum local

v0d1ch committed:
Rename headResolvedInput to healthyHeadTxOut

ffakenz committed:
do not limit to a single token value

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
update comment on commit mutation

v0d1ch committed:
Extract hydraHeadV1 and hasST into separate module

v0d1ch committed:
Add CurrencySymbol to the commit datum

v0d1ch committed:
Check ST is present in the CollectCom commit output

v0d1ch committed:
Remove traces

v0d1ch committed:
Small refactor

v0d1ch committed:
Remove a solved fixme

v0d1ch committed:
More PR review comments

v0d1ch committed:
Check for PTs burning and headId in v_commit

v0d1ch committed:
Comment out branch of utxo generation to make the tests pass

v0d1ch committed:
Remove the redundant txspec test

v0d1ch committed:
Add the ST check to v_initial

v0d1ch committed:
Change mustBurnPTs to check ST removal

v0d1ch committed:
Remove the redundant v_head checks from v_commit

v0d1ch committed:
Remove redundant imports

v0d1ch committed:
Remove collectCom tx-spec test

v0d1ch committed:
Remove MutateHeadScriptInput since it is not relevant anymore for v_commit

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -161,13 +164,14 @@ mkInitialOutput networkId tokenPolicyId (verificationKeyHash -> pkh) =
   initialScript =
     fromPlutusScript Initial.validatorScript
   initialDatum =
-    mkTxOutDatum $ Initial.datum ()
```
Should: not break with the convention of having a `Module.datum` smart constructor for suitable datum.

Also.. I'm not sure if we want to introduce the `newtype InitialDatum` as it does not give us any more type safety (see this piece of code here) and if I'm not mistaken, the plutus compiler does NOT strip off the constructor, so we have an additional layer of wrapping on-chain as well.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -577,7 +594,7 @@ observeInitTx ::
   -- FIXME: This is affected by "same structure datum attacks", we should be
   -- using the Head script address instead.
-  (ix, headOut, headData, Head.Initial cp ps) <- findFirst headOutput indexedOutputs
+  (ix, headOut, headData, Head.Initial cp ps _headPolicyId) <- findFirst headOutput indexedOutputs
```
This code should also check that the policy id in the datum is consistent with the one used in the value (e.g. the ST in the first output) to qualify a correct initTx.

We do not need to do it now, but maybe add a FIXME about this to the one above?

ch1bo replied:
Yeah I think this is exactly the issue. The fact that nothing on-chain check whether the datum in the outputs is having the same `HeadId` than the policy of the tokens distributed means that the off-chain needs to decide to ignore such a flawed head.

Alternatively we can put those checks into the minting policy, but that would be just wasteful and increase on-chain code surface.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Abort.hs:
```diff
@@ -143,16 +140,15 @@ data AbortMutation
     , SomeMutation DropOneCommitOutput
         . RemoveOutput
         <$> choose (0, fromIntegral (length (txOuts' tx) - 1))
-    , SomeMutation MutateHeadScriptInput <$> (ChangeInput (headTxIn utxo) <$> anyPayToPubKeyTxOut <*> pure Nothing)
```
What's the rationale of removing this? (I do not see much value in this mutation as well, but want to check)

ch1bo committed:
Rename variables in Close mutation test

We use the healthyXXX prefix for our starting values to identify them as
good values to mutate "away" from.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -191,6 +201,16 @@ genCloseMutation (tx, _utxo) =
+            [ ChangeOutput 0 (replacePolicyIdWith Fixture.testPolicyId otherHeadId headTxOut)
+            , ChangeInput
+                headInput
+                (toUTxOContext $ replacePolicyIdWith Fixture.testPolicyId otherHeadId (toTxContext headResolvedInput))
```
Unneeded conversion from/to `TxContext`. Dropped this in 5c9c9a1f8de213db84cd1612deebe0783b3a24d0

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -182,24 +128,20 @@ checkAbort context@ScriptContext{scriptContextTxInfo = txInfo} headContext parti
-  HeadContext ->
   -- | Initial state
-  (ContestationPeriod, [Party]) ->
+  State ->
```
This is a step back. It's less type safe. If you want to get rid of large tuples, we should introduce also on-chain the `InitialState` sub-type, i.e. ` data State = ... | Initial InitialState | ... `

ch1bo committed:
Fix haddock mentions of ST/PT

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -171,6 +115,8 @@ checkAbort context@ScriptContext{scriptContextTxInfo = txInfo} headContext parti
 --
 --   * The transaction is performed (i.e. signed) by one of the head participants
 --
+--   * ST token is present in the output
```
ST already means token. I updated the description of all PT / ST mentions in the code in  82c0e75a1bc11d20178d639c32b95d42288796ed

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -263,22 +206,28 @@ checkCollectCom context@ScriptContext{scriptContextTxInfo = txInfo} headContext
+    && mustBeSignedByParticipant ctx headPolicyId
+    && hasST headPolicyId outValue
  where
-  hasBoundedValidity = traceIfFalse "hasBoundedValidity check failed" $ tMax - tMin <= cp
```
We still miss this `traceIfFalse`

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -343,27 +288,60 @@ checkContest ::
 
   mustBeWithinContestationPeriod =
     case ivTo (txInfoValidRange scriptContextTxInfo) of
-      UpperBound (Finite time) _ -> traceIfFalse "upper bound validity beyond contestation deadline" $ time <= contestationDeadline
```
Also this one was helpful.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -343,27 +288,60 @@ checkContest ::
+  outValue =
+    maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
   mustBeNewer =
-    traceIfFalse "too old snapshot" $
```
Also this one was helpful.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -372,10 +350,9 @@ checkHeadOutputDatum ctx d =
-        Just actualHash == expectedHash
+      Just actualHash == expectedHash
     OutputDatum actual ->
-      traceIfFalse "output datum mismatch" $ getDatum actual == expectedData
```
Also these traces were helpful.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -343,27 +288,60 @@ checkContest ::
+
+  afterContestationDeadline =
+    case ivFrom (txInfoValidRange txInfo) of
+      LowerBound (Finite time) _ -> time > contestationDeadline
```
Missing trace.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -343,27 +288,60 @@ checkContest ::
+checkFanout utxoHash contestationDeadline numberOfFanoutOutputs ScriptContext{scriptContextTxInfo = txInfo} =
+  hasSameUTxOHash && afterContestationDeadline
+ where
+  hasSameUTxOHash = fannedOutUtxoHash == utxoHash
```
Missing trace.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -502,15 +471,13 @@ hashTxOuts =
 
 verifySnapshotSignature :: [Party] -> SnapshotNumber -> BuiltinByteString -> [Signature] -> Bool
 verifySnapshotSignature parties snapshotNumber utxoHash sigs =
-  traceIfFalse "signature verification failed" $
```
Also this trace was helpful.

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -502,15 +471,13 @@ hashTxOuts =
 
 verifyPartySignature :: SnapshotNumber -> BuiltinByteString -> Party -> Signature -> Bool
-verifyPartySignature snapshotNumber utxoHash party signed =
-  traceIfFalse "party signature verification failed" $
```
Also this trace was helpful.

ch1bo commented on hydra-plutus/exe/inspect-script/Main.hs:
```diff
@@ -102,7 +103,7 @@ main = do
     ]
 
-  headDatum = toData $ Head.Initial 1_000_000_000_000 []
+  headDatum = toData $ Head.Initial 1_000_000_000_000 [] (CurrencySymbol hydraHeadV1)
```
This is not what the policy ID will be. It's rather a `hash(mu_head(someTxIn))`.

I think we should not bother and just drop the datum serialization here. It's just examples.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -279,7 +192,8 @@ generateCommitUTxOs parties = do
     commitScript = fromPlutusScript Commit.validatorScript
-    commitDatum = mkCommitDatum party Head.validatorHash utxo
+    commitDatum (Just (_input, _)) = mkCommitDatum party Head.validatorHash utxo (toPlutusCurrencySymbol testPolicyId)
+    commitDatum Nothing = error "Missing utxo"
```
Could: follow the style guide and use lambda case (and a new line between where clauses)

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/StateSpec.hs:
```diff
@@ -333,8 +334,7 @@ forAllClose ::
   (UTxO -> Tx -> property) ->
   Property
 forAllClose action = do
-  -- FIXME: we should not hardcode number of parties but generate it within bounds
```
It's still hard-coded. Just a different value. What this FIXME says is that we should generate between `1..maximumNumberOfParties`

ch1bo replied:
Same holds for these other generators. They should try most possible combinations / be "as arbitrary as possible".

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -121,7 +120,6 @@
 --       30.5% MutateOpenUtxoHash
 --       27.0% MutateHeadTransition
 --       23.5% MutateOpenOutputValue
---       19.0% MutateHeadScriptInput
```
This is not 100% now .. keep it, it's outdated anyways.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -193,7 +191,9 @@ propTransactionDoesNotValidate (tx, lookupUTxO) =
-          any isLeft (Map.elems redeemerReport)
+          any
+            isLeft
+            (Map.elems redeemerReport)
```
Formatter issues? In my case it would stay one line.

ch1bo committed:
Use a dedicatd TxIn for Contest mutation tests

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -50,27 +51,27 @@ healthyContestTx =
   closedThreadOutput =
     ClosedThreadOutput
-      { closedThreadUTxO = (headInput, headResolvedInput, headDatum)
+      { closedThreadUTxO = (testSeedInput, healthyHeadTxOut, headDatum)
```
Changed this to use a dedicated `TxIn` in fbd2da5ae02709fb9745cf7c901b923554878a39. It has nothing to do with the seed input after all.

ch1bo committed:
Follow naming scheme in Commit mutation test

ch1bo reviewed (changes requested):
We are still missing a lot of error traces in the scripts.

This is required before I can approve.

I already addressed some of the inconsistencies and oversights. In general, the PR really shows that it's open too long already. Messy.

v0d1ch committed:
Use type alias for v_initial script datum

pgrange committed:
Using lambda case style guide

pgrange committed:
We did'nt fix this acutally

pgrange committed:
FIX comment

v0d1ch committed:
Don't use InitialState for checkCollectCom

v0d1ch committed:
Add back the traces in the v_head

v0d1ch committed:
Rename State head fields to be uniform

abailly committed:
Output a realistic head datum in inspect-script

Also refactor couple functions to move them to hydra-plutus. This
change is larger than it should be.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -194,7 +190,7 @@ commitTx scriptRegistry networkId party utxo (initialInput, out, vkh) =
     fst (initialReference scriptRegistry)
   initialDatum =
-    mkScriptDatum $ Initial.datum ()
+    mkScriptDatum $ Initial.datum (headIdToCurrencySymbol headId)
```
:+1:

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -53,107 +54,50 @@ import Plutus.V1.Ledger.Value (assetClass, assetClassValue, valueOf)
-  HeadContext{headCurrencySymbol} = headContext
-
   mustBurnAllHeadTokens =
-    traceIfFalse "number of inputs do not match number of parties" $
```
:(

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -182,24 +128,19 @@ checkAbort context@ScriptContext{scriptContextTxInfo = txInfo} headContext parti
+  outValue =
+    maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
   everyoneHasCommitted =
-    traceIfFalse "not everyone committed" $
```
:(

ch1bo commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -372,10 +350,10 @@ checkHeadOutputDatum ctx d =
     NoOutputDatum ->
       traceError "missing datum"
     OutputDatumHash actualHash ->
-      traceIfFalse "output datum hash mismatch" $
```
:(

ch1bo committed:
Make import of orphan isntance more obvious

Hydra.Chain.Direct.Tx was not directly including the required orphan
instance, but Hydra.Ledger.Cardano is.

ch1bo committed:
Re-add traces

ch1bo reviewed (approved):
Re-added some last traces in da1cdd92cfdb37cedc929a9e39c3705e4dd989cf.. now it's good to go :)

v0d1ch commented:
> Re-added some last traces in [da1cdd9](https://github.com/input-output-hk/hydra/commit/da1cdd92cfdb37cedc929a9e39c3705e4dd989cf).. now it's good to go :)

Nice but now our tx sizes grew bigger :) Shall I increase the limit or revert?

v0d1ch committed:
Reduce trace messages length

v0d1ch commented:
We reduced the tx traces and hitting the button now!
